pkg/engine/compat.go

```diff
@@ -33,119 +33,192 @@ var (
+	// After all the values are collected and converted we transform the columnar representation to a row-based one.
+
+	b.ensureRowBuilders(numRows)
+
+	// Convert arrow values to our format column by column
 	for colIdx := range int(rec.NumCols()) {
 		col := rec.Column(colIdx)
-		// Ignore column values that are NULL or invalid
-		if col.IsNull(i) || !col.IsValid(i) {
-			continue
-		}
 
 		field := rec.Schema().Field(colIdx)
 		ident, err := semconv.ParseFQN(field.Name)
 		if err != nil {
 			continue
 		}
-
 		shortName := ident.ShortName()
 
-		// Extract line
-		if ident.Equal(semconv.ColumnIdentMessage) {
-			entry.Line = col.(*array.String).Value(i)
-			continue
+		switch true {
+
+		// Log line
+		case ident.Equal(semconv.ColumnIdentMessage):
+			lineCol := col.(*array.String)
+			forEachNotNullRowColValue(numRows, lineCol, func(rowIdx int) {
+				b.rowBuilders[rowIdx].line = lineCol.Value(rowIdx)
+			})
+
+		// Timestamp
+		case ident.Equal(semconv.ColumnIdentTimestamp):
+			tsCol := col.(*array.Timestamp)
+			forEachNotNullRowColValue(numRows, tsCol, func(rowIdx int) {
+				b.rowBuilders[rowIdx].timestamp = time.Unix(0, int64(tsCol.Value(rowIdx)))
+			})
+
+		// One of the label columns
+		case ident.ColumnType() == types.ColumnTypeLabel:
+			labelCol := col.(*array.String)
+			forEachNotNullRowColValue(numRows, labelCol, func(rowIdx int) {
+				b.rowBuilders[rowIdx].lbsBuilder.Set(shortName, labelCol.Value(rowIdx))
+			})
+
+		// One of the metadata columns
+		case ident.ColumnType() == types.ColumnTypeMetadata:
+			metadataCol := col.(*array.String)
+			forEachNotNullRowColValue(numRows, metadataCol, func(rowIdx int) {
+				val := metadataCol.Value(rowIdx)
+				b.rowBuilders[rowIdx].metadataBuilder.Set(shortName, val)
+				b.rowBuilders[rowIdx].lbsBuilder.Set(shortName, val)
+			})
+
+		// One of the parsed columns
+		case ident.ColumnType() == types.ColumnTypeParsed:
+			parsedCol := col.(*array.String)
+
+			// TODO: keep errors if --strict is set
+			// These are reserved column names used to track parsing errors. We are dropping them until
+			// we add support for --strict parsing.
+			if shortName == types.ColumnNameError || shortName == types.ColumnNameErrorDetails {
+				continue
+			}
+
+			forEachNotNullRowColValue(numRows, parsedCol, func(rowIdx int) {
+				parsedVal := parsedCol.Value(rowIdx)
+				if b.rowBuilders[rowIdx].parsedBuilder.Get(shortName) != "" {
+					return
+				}
+				b.rowBuilders[rowIdx].parsedBuilder.Set(shortName, parsedVal)
+				b.rowBuilders[rowIdx].lbsBuilder.Set(shortName, parsedVal)
+				if b.rowBuilders[rowIdx].metadataBuilder.Get(shortName) != "" {
+					b.rowBuilders[rowIdx].metadataBuilder.Del(shortName)
+				}
+			})
 		}
+	}
 
-		// Extract timestamp
-		if ident.Equal(semconv.ColumnIdentTimestamp) {
-			entry.Timestamp = time.Unix(0, int64(col.(*array.Timestamp).Value(i)))
+	// Convert columnar representation to a row-based one
+	for rowIdx := 0; rowIdx < numRows; rowIdx++ {
```

Comment:
Could me written as `for rowIdx := range numRows {`

Reply:
Thanks, indeed 😄 Fixed